tests/test_async.py

```diff
@@ -0,0 +1,276 @@
+    async with registry as r:
+        my_counter = await r.aget(MyAsyncAPIContextCounter)
+        assert my_counter.entered_context_counter == 1
+        await r.aget(MyAsyncApi)
+        assert my_counter.entered_context_counter == 1
+    assert my_counter.exited_context_counter == 1
+
+
+async def test_async_context_outside_context_manager(registry: Registry) -> None:
+    with pytest.raises(AssertionError):
+        # attempting to instantiate a class
+        # marked with @async_context without
+        # being in an async context should
+        # raise an error
+        _ = await registry.aget(MyAsyncApi)
+
+
+async def test_try_instantiate_async_class_with_sync_api(registry: Registry) -> None:
+    with pytest.raises(AssertionError):
+        # attempting to instantiate a class
+        # marked with @async_context using sync API
+        # should raise an error
+        _ = registry[MyDependencyAsync]
+
+    with pytest.raises(AssertionError):
+        # still throws an error even when registry context
+        # has been entered
+        async with registry as r:
+            _ = r[MyAsyncApi]
+
+
+async def test_context_manager_aenter_must_return_self(registry: Registry) -> None:
+    """
+    Async context manager must return self from aenter method,
+    throw value error otherwise.
+    """
+    async with registry as r:
+        with pytest.raises(ValueError):
+            _ = await r.aget(BadContextManager)
+
+
+async def test_config_in_async(registry: Registry) -> None:
+    async with registry as r:
+        r = await r.aget(MyAsyncApiWithConfig)
+        assert r.nested == "c"
+        assert r.flat == 2
+
+
+async def test_entering_already_entered_registry_throws(registry: Registry) -> None:
+    async with registry as r:
+        with pytest.raises(AssertionError):
+            async with r:
+                pass
+
+
+async def test_define(registry: Registry) -> None:
+    async with registry as r:
+        my_api = await r.aget(MY_ASYNC_API_DEFINE)
+        assert my_api.text == TEXT
+        assert my_api.in_context == True
+        assert my_api.dep_async.in_context == True
+        assert my_api.dep_not_specified.in_context == False
+        assert my_api.dep_context_counter.in_context == True
+        assert my_api.dep_context_counter.entered_context_counter == 1
+
+    assert my_api.in_context == False
+    assert my_api.dep_async.in_context == False
+    assert my_api.dep_not_specified.in_context == False
+    assert my_api.dep_context_counter.in_context == False
+    assert my_api.dep_context_counter.exited_context_counter == 1
+
+
+def test_get_item_sync_class_async_dependency_throws(registry: Registry) -> None:
+    with pytest.raises(AssertionError):
+        _ = registry[MySyncClassWithAsyncDependency]
+
+
+def test_get_sync_class_async_dependency_throws(registry: Registry) -> None:
+    with pytest.raises(AssertionError):
+        _ = registry.get(MySyncClassWithAsyncDependency, AUTO_OR_NONE)
```

Comment:
The API for `registry.get` is a bit odd, you need to specify `AUTO_OR_NONE` to get the `Registry` instance to actually attempt to create your class instance (see [here](https://github.com/duolingo/minject/blob/cb530aae9cacf86fd411324e46e91484759e1573/minject/registry.py#L194-L199)). I think it makes sense to remove `get` from the public API for registry, and only push people towards using `Registry.__getitem__`.

Any objection to this @mmchenry-duolingo?

Reply:
yes, get is probably unnecessary